Integrate USEEIO import factors code with CEDA

cclin130 opened the pull request:
This PR integrates CEDA data artifacts with the USEEIO import factor code to generate CEDA import factors. This PR can be read commit-by-commit. 

1. The first commit adds CEDA concordance files, country names, and appends the CEDA configuration to `mrio_config.yml`. Of note, we add a new section in the MRIO config called `calculation_configs` for branching logic in our method. I added a `use_industry_output_for_usa_electricity_imports` flag because Exiobase uses industry output electricity for the US (as discussed here: https://github.com/USEPA/USEEIO/discussions/41 ) while we do not think CEDA needs this. This is pending validation of the CEDA import factor data artifacts.

2. The second commit modifies `generate_import_factors.py` to work with the CEDA MRIO config. Specifically, it refactors some lingering Exiobase-specific logic. It adds `exiobase_helpers.py` and `ceda_helpers.py` files for storing MRIO-specific processing/cleaning logic.


## Testing
I ran the script to generate CEDA import factors for all years. I also re-ran the Exiobase code for 2019 and confirmed the values are roughly the same as what's uploaded online (with some rounding error -- let me if this should not be the case)

WesIngwersen commented on import_emission_factors/generate_import_factors.py:
This comment is not specific to this PR but rather the extension of the existing code makes this shortcoming more clear:

This modules imports modules for MRIO sources and contains code forks - like IF statements - for handling specific sources as well. This means this module has to be updated each time a new source is being handled. This is not the best design. 

We have code among the [TIEM](https://www.epa.gov/land-research/tools-industrial-ecology-modeling-tiem) that can provide an example of a superior module design. See for instance FLOWSA's [generateflowbyactivity.py](https://github.com/USEPA/flowsa/blob/master/flowsa/generateflowbyactivity.py).   Code that is required to parse and further transform a specific source is put into a separate source module.  Some of that is describing in [this Wiki section](https://github.com/USEPA/flowsa/wiki/Contributing#creating-a-flow-by-activity-dataset). 

Files here such as ceda_helpers.py and the yaml configs partially implement but do not fully implement this option. I would recommend further refactoring to mature the code to eliminate specific MRIO imports and code forks from this file.

bl-young replied:
Yes agreed - my initial effort to make the code more general was a bit rushed, but I agree this is the way to go.

I may review this code as is and make this update after pulling in this PR, to avoid substantial conflicts, but I will assess after I look closer.

cclin130 replied:
@bl-young sounds good, I'm also happy to help with this refactor if you'd like. @WesIngwersen thank you for the resources, I will also take a look for my own learning :)

bl-young reviewed (changes requested):
Two minor changes, I was able to run the 2019 factors with the pkl files you shared. Will you be providing a script as well that would enable generation of those pkl files based on raw CEDA data (not available to most users, I expect)? (Similar to download_exiobase.py)

bl-young commented:
> I recall in our live sync, @WesIngwersen said that it was OK for us to not provide users access to the raw CEDA data (this is why I have shared them via gdrive) and the code by itself would still be useful information to users. Watershed is looking into releasing a free tier of CEDA, but we don't want to postpone the release of this import factors work for it because work for the former is still in its infancy.

No, I agree. No need to provide the raw data. I guess what I was asking is, I assume there was a transformation of the raw data to the pickle file. There is some value in showing how that was done, though I don't know what the raw CEDA data looks like so maybe it is very straightforward. But we can discuss further.

cclin130 commented:
@bl-young understood. I did do some transformation, yes. I did this in our Watershed repository because I used some internal helpers that we have to fetch assets from our servers, etc. Happy to discuss further, perhaps on Friday? And if we want to share that code I can add it as a follow up PR

cclin130 commented:
@bl-young addressed comments!

cclin130 commented:
@bl-young two more additions for your review in the latest commit:

1. I added our scripts used to process raw CEDA files for use. @MoLi7 Please also take a look as well. The code requires the user to already have these files locally, I have made them available via gdrive under the `Data/mrio_models` folder
2. I added a config param to specify price type because I noticed we always use `Basic` but CEDA is in producer price.

bl-young commented:
> 1. I added our scripts used to process raw CEDA files for use. @MoLi7 Please also take a look as well. The code requires the user to already have these files locally, I have made them available via gdrive under the `Data/mrio_models` folder

Great - thanks for sharing! I will try to give it a test before Friday, but thanks for sharing for transparency!

> 2. I added a config param to specify price type because I noticed we always use `Basic` but CEDA is in producer price.

Perfect - yes if they are already in producer price than we don't need to do a [price type conversion in useeior](https://github.com/USEPA/useeior/blob/master/R/ExternalImportFactors.R#L48-L59). Good catch.

bl-young commented:
I will pull these into the ceda branch pending any final updates/revisions

WesIngwersen commented on import_emission_factors/generate_import_factors.py:
```diff
@@ -77,15 +76,19 @@ def generate_import_emission_factors(years: list, schema=2012, calc_tiva=False):
         bilateral = pull_mrio_data(year, opt = "bilateral")
         output = pull_mrio_data(year, opt = "output")
         export_field = list(config.get('exports').values())[0]
+
```
The import shares should be independent of the MRIO. I just created this issue so we can work on that.   #53
